examples/ex3p.cpp

```diff
@@ -181,26 +186,44 @@ int main(int argc, char *argv[])
    if (static_cond) { a->EnableStaticCondensation(); }
    a->Assemble();
 
-   HypreParMatrix A;
+   OperatorPtr A;
    Vector B, X;
    a->FormLinearSystem(ess_tdof_list, x, *b, A, X, B);
 
-   if (myid == 0)
+   // 12. Define and apply a parallel PCG solver for AX=B with the AMS
+   //     preconditioner from hypre, in the full assembly case.
+   //     With partial assembly, use no preconditioner, for now.
+
+   if (pa)
    {
-      cout << "Size of linear system: " << A.GetGlobalNumRows() << endl;
+      Vector tdiag_pa(fespace->GetTrueVSize());
+      a->AssembleDiagonal(tdiag_pa);
+
+      OperatorJacobiSmoother Jacobi(tdiag_pa, ess_tdof_list, 1.0);
+
+      CGSolver cg(MPI_COMM_WORLD);
+      cg.SetRelTol(1e-12);
+      cg.SetMaxIter(1000);
+      cg.SetPrintLevel(1);
+      cg.SetOperator(*A);
+      cg.SetPreconditioner(Jacobi);
+      cg.Mult(B, X);
    }
+   else
+   {
+      ParFiniteElementSpace *prec_fespace =
+         (a->StaticCondensationIsEnabled() ? a->SCParFESpace() : fespace);
+      HypreSolver *ams = new HypreAMS(*A.As<HypreParMatrix>(), prec_fespace);
```

Comment:
To be more uniform can we make `ams` and `pcg` objects rather than pointers like `Jacobi` and `cg` in the earlier code?